GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1157,12 +1157,35 @@ private void UpdatePlaceholders()
                     //  1. Ensures child folders are deleted before their parents
                     //  2. Ensures that folders that have been deleted by git (but are still in the projection) are found before their
                     //     parent folder is re-expanded (only applies on platforms where EnumerationExpandsDirectories is true)
+                    Dictionary<string, long> availableSizes = new Dictionary<string, long>();
                     foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderFoldersListCopy.OrderByDescending(x => x.Path))
                     {
-                        // Remove folder placeholders before re-expansion to ensure that projection changes that convert a folder to a file work
-                        // properly
-                        if (!this.RemoveFolderPlaceholderIfEmpty(folderPlaceholder, addPlaceholderToUpdatedPlaceholders, folderPlaceholdersToKeep))
+                        bool keepFolder = true;
+                        if (!folderPlaceholdersToKeep.Contains(folderPlaceholder.Path))
                         {
+                            this.GetChildNameAndParentKey(folderPlaceholder.Path, out string childName, out string parentKey);
+                            FolderData folderData = this.GetProjectedFolderEntryData(CancellationToken.None, blobSizesConnection, availableSizes, childName, parentKey, out string _) as FolderData;
+
+                            // Check the projection for the folder to determine if the folder needs to be deleted
+                            // The delete will be attempted if one of the following is true
+                            // 1. not in the projection anymore
+                            // 2. in the projection but is not a folder in the projection
+                            // 3. Folder no longer exist on disk - this is to remove the tombstone
+                            if (folderData == null ||
+                                !folderData.IsFolder ||
+                                !this.context.FileSystem.DirectoryExists(Path.Combine(this.context.Enlistment.WorkingDirectoryRoot, folderPlaceholder.Path)))
+                            {
+                                keepFolder = !this.RemoveFolderPlaceholderIfEmpty(folderPlaceholder);
+                                ++deleteFolderPlaceholderAttempted;
+                            }

```

Comment:
If `keepFolder` is true (after making it to this point), should we call `AddParentFoldersToListToKeep` for the parent(s) of the folder?

I think that would save us from projection+disk lookups for all ancestors of the folder (e.g. in the scenarios where there are a lot of folder placeholders but not many file placeholders).

Reply:
I don't think that would hurt anything.

Reply:
Do you think there would be a benefit? 

(I thought maybe there would be, but was hoping you'd double check my thoughts).

I was thinking if there was a directory tree like this with no file placeholders in it:

`src\a\b\c\d\e\f\g`

The current code would check each of these directories (assuming it decides to keep g):

`src\a\b\c\d\e\f\g`
`src\a\b\c\d\e\f`
`src\a\b\c\d\e`
`src\a\b\c\d`
`src\a\b\c`
`src\a\b`
`src\a`

And that we could skip all those checks by calling `AddParentFoldersToListToKeep`.

Let me know what you think.

Reply:
Okay let me think about the scenarios.  The code gets here and `keepFolder` is `true` when it is still in the projection and it is a folder and it exists on disk.  If any of those are not the case we will try to delete the folder and if we get an error or directory not empty when trying to delete, `keepFolder` will be true.

1. folder no longer in the projection
Attempt the delete
2. folder still in the projection as a folder and on disk
Can keep all parents
3. folder still in the projection but doesn't exist on disk
Attempt the delete
4. folder path still in the projection as a file
Attempt the delete

Attempting to delete the folder:
1. deleting folder gets directory not empty - `keepFolder` will be `true`
Can keep all parents because they will get not empty as well.
2. deleting folder gets error - `keepFolder` will be `true`
Keep the folder placeholder and assume deletes on parents will fail as well so skip them
3. deleting folder gets success or not found - `keepFolder` will be `false`
Need to continue deleting parents

So from my thinking we would be okay.